{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(dplyr)\n",
    "library(plyr)\n",
    "library(data.table)\n",
    "library(readr)\n",
    "library(magrittr)\n",
    "\n",
    "# Import the output files from RNA Mutect\n",
    "data_all <- list.files(path = \"~/R/ltrc_rna_samples_dir\",  # Identify all patient files\n",
    "                       pattern = \"*.maf.annotated\", full.names = TRUE) %>% \n",
    "  lapply(read_table) %>%                              # Store all files in list\n",
    "  bind_rows                                         # Combine data sets into one data set \n",
    "data_all     \n",
    "#save data\n",
    "write.table(data_all,\"~/R/LTRC_data/total_rna_data\") # 21,363,023 variants\n",
    "\n",
    "\n",
    "## filter tumor fractions to 10-30%\n",
    "#import above data\n",
    "total_rna_variants <- read.table(\"~/R/LTRC_data/total_rna_data\", header = TRUE)\n",
    "filtered_rna_variants <- total_rna_variants[total_rna_variants$tumor_f > 0.1 & total_rna_variants$tumor_f < 0.3, ]\n",
    "#save data\n",
    "write.table(filtered_rna_variants,\"~/R/LTRC_data/filtered_rna_data\") # 1,181,035 variants, 1346 patients\n",
    "\n",
    "\n",
    "#setting unique key function to exclude any duplicate variants\n",
    "#import above data\n",
    "filtered_rna_variants <- read.table(\"~/R/LTRC_data/filtered_rna_data\", header = TRUE)\n",
    "unique_rna_variants <- filtered_rna_variants[!duplicated(filtered_rna_variants[2:7]),]\n",
    "write.table(unique_rna_variants,\"~/R/LTRC_data/unique_rna_variants\") #134,540 variants\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Convert file format to maf for future maf22vcf conversion\n",
    "```{r}\n",
    "\n",
    "filtered_rna_variants <- read.table(\"~/R/LTRC_data/filtered_rna_data\", header = TRUE)\n",
    "#\n",
    "colnames(filtered_rna_variants)[1] <- \"NCBI_Build\"\n",
    "filtered_rna_variants$NCBI_Build <- \"GRCh38\"\n",
    "#\n",
    "colnames(filtered_rna_variants)[2] <- \"Chromosome\"\n",
    "filtered_rna_variants$Chromosome <- sub(\"chr\", \"\", as.character(filtered_rna_variants$Chromosome))\n",
    "#\n",
    "colnames(filtered_rna_variants)[8] <- \"Tumor_Sample_Barcode\"\n",
    "filtered_rna_variants$Tumor_Sample_Barcode <- \"Tumor\"\n",
    "#\n",
    "colnames(filtered_rna_variants)[3] <- \"Start_Position\"\n",
    "#\n",
    "colnames(filtered_rna_variants)[4] <- \"End_Position\"\n",
    "#\n",
    "colnames(filtered_rna_variants)[5] <- \"Reference_Allele\"\n",
    "#\n",
    "colnames(filtered_rna_variants)[6] <- \"Tumor_Seq_Allele1\"\n",
    "#\n",
    "colnames(filtered_rna_variants)[7] <- \"Tumor_Seq_Allele2\"\n",
    "#\n",
    "filtered_rna_variants$Entrez_Gene_Id <- 0\n",
    "filtered_rna_variants$Center <- \".\"\n",
    "filtered_rna_variants$Strand <- \".\"\n",
    "filtered_rna_variants$Variant_Classification <- \".\"\n",
    "filtered_rna_variants$Variant_Type <- \".\"\n",
    "filtered_rna_variants$dbSNP_RS <- \".\"\n",
    "filtered_rna_variants$dbSNP_Val_Status <- \".\"\n",
    "filtered_rna_variants$Hugo_Symbol <- \"\"\n",
    "filtered_rna_variants$dbSNP_RS <- \".\"\n",
    "filtered_rna_variants$dbSNP_Val_Status <- \".\"\n",
    "#\n",
    "filtered_rna_variants$Matched_Norm_Sample_Barcode <- \"\"\n",
    "filtered_rna_variants$Match_Norm_Seq_Allele1 <- \"\"\n",
    "filtered_rna_variants$Match_Norm_Seq_Allele2 <- \"\"\n",
    "filtered_rna_variants$Tumor_Validation_Allele1 <- \"\"\n",
    "filtered_rna_variants$Tumor_Validation_Allele2 <- \"\"\n",
    "filtered_rna_variants$Match_Norm_Validation_Allele1 <- \"\"\n",
    "filtered_rna_variants$Match_Norm_Validation_Allele2 <- \"\"\n",
    "filtered_rna_variants$Verification_Status <- \"\"\n",
    "filtered_rna_variants$Validation_Status <- \"\"\n",
    "filtered_rna_variants$Mutation_Status <- \"\"\n",
    "filtered_rna_variants$Sequencing_Phase <- \"\"\n",
    "filtered_rna_variants$Sequence_Source <- \"\"\n",
    "filtered_rna_variants$Validation_Method <- \"\"\n",
    "filtered_rna_variants$Score <- \"\"\n",
    "filtered_rna_variants$BAM_File <- \"\"\n",
    "filtered_rna_variants$Sequencer <- \"\"\n",
    "filtered_rna_variants$Tumor_Sample_UUID <- \"\"\n",
    "filtered_rna_variants$Matched_Norm_Sample_UUID <- \"\"\n",
    "#\n",
    "filtered_rna_variants <- filtered_rna_variants[, c(\"Hugo_Symbol\", \"Entrez_Gene_Id\", \"Center\", \"NCBI_Build\",\"Chromosome\", \"Start_Position\", \"End_Position\", \"Strand\", \"Variant_Classification\", \"Variant_Type\", \"Reference_Allele\", \"Tumor_Seq_Allele1\", \"Tumor_Seq_Allele2\", \"dbSNP_RS\", \"dbSNP_Val_Status\", \"Tumor_Sample_Barcode\", \"Matched_Norm_Sample_Barcode\", \"Match_Norm_Seq_Allele1\", \"Match_Norm_Seq_Allele2\", \"Tumor_Validation_Allele1\", \"Tumor_Validation_Allele2\", \"Match_Norm_Validation_Allele1\", \"Match_Norm_Validation_Allele2\", \"Verification_Status\", \"Validation_Status\", \"Mutation_Status\", \"Sequencing_Phase\", \"Sequence_Source\", \"Validation_Method\", \"Score\", \"BAM_File\", \"Sequencer\", \"Tumor_Sample_UUID\", \"Matched_Norm_Sample_UUID\", \"tumor_f\", \"t_ref_count\", \"t_alt_count\", \"init_t_lod\", \"t_lod_fstar\")]\n",
    "write.table(filtered_rna_variants, file = \"~/R/LTRC_data/filtered_rna_variants.maf\", sep = \"\\t\", quote=FALSE, row.names=FALSE )\n",
    "\n",
    "\n",
    "#filtered_rna_variants_maf <- read.delim(\"~/R/LTRC_data/filtered_rna_variants.maf\", sep = \"\\t\", header = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Annotate Datasets\n",
    "#This section requires access to a Linux terminal (Terra terminal)\n",
    "\n",
    "\n",
    "#Install Ensembl's VEP:\n",
    "#https://useast.ensembl.org/info/docs/tools/vep/script/vep_download.html#installer\n",
    "\n",
    "#if installing dependencies (GRCh38) is not successful, do it manually:\n",
    "# VEP's offline cache for GRCh38\n",
    "mkdir -p $HOME/.vep/homo_sapiens/102_GRCh38/\n",
    "rsync -avr --progress rsync://ftp.ensembl.org/ensembl/pub/release-106/variation/indexed_vep_cache/homo_sapiens_vep_106_GRCh38.tar.gz $HOME/.vep/\n",
    "tar -zxf $HOME/.vep/homo_sapiens_vep_106_GRCh38.tar.gz -C $HOME/.vep/\n",
    "\n",
    "# reference FASTA file\n",
    "rsync -avr --progress rsync://ftp.ensembl.org/ensembl/pub/release-106/fasta/homo_sapiens/dna_index/ $HOME/.vep/homo_sapiens/106_GRCh38/\n",
    "\n",
    "# if FTP site is not working, try: \n",
    "curl -O ftp://ftp.ensembl.org/pub/release-98/variation/indexed_vep_cache/homo_sapiens_vep_98_GRCh38.tar.gz # or any link you need from above\n",
    "\n",
    "\n",
    "\n",
    "#conversion maf2vcf\n",
    "#1. perl -lne 's/\\r//; print \"$_\";' samples.maf > samples_fixed.maf (Needed to run if CR uses line breaks)\n",
    "#2. perl maf2vcf_1.pl --input-maf unique_fixed.maf --output-dir ~/snpEff/vcf_files/ --ref-fasta ~/.vep/homo_sapiens/105_GRCh38/Homo_sapiens.GRCh38.dna.toplevel.fa.gz\n",
    "\n",
    "\n",
    "#Annotate using SNPsift\n",
    "#java -jar SnpSift.jar annotate hg38 variants.vcf > variants_annotated.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Make annotated data more presentable \n",
    "\n",
    "library(data.table)\n",
    "filtered_ann_variants <- fread(\"~/R/LTRC_data/filtered_rna_rna_variants_annotated.vcf\")\n",
    "\n",
    "\n",
    "#separate INFO section to extrac Gene info\n",
    "library(dplyr)\n",
    "gene <- data.frame(filtered_ann_variants[[8]]%>%strsplit(\"|\", fixed=T)%>%sapply(\"[\", 4))\n",
    "mutation <- data.frame(filtered_ann_variants[[8]]%>%strsplit(\"|\", fixed=T)%>%sapply(\"[\", 2))\n",
    "\n",
    "variants <- cbind(filtered_ann_variants,gene,mutation)\n",
    "colnames(variants)[12] <- \"GENE\"\n",
    "colnames(variants)[13] <- \"MUTATION\"\n",
    "variants$NORMAL <- NULL\n",
    "\n",
    "#rearrange columns\n",
    "variants <- variants[, c(1:7,11,12,8:10)]\n",
    "write.table(variants, file = \"~/R/LTRC_data/variants\", sep = \"\\t\", quote=FALSE, row.names=FALSE )\n",
    "\n",
    "\n",
    "#dataset needed to match the patient ID\n",
    "filtered_test <- fread(\"~/R/LTRC_data/filtered_test.maf\")\n",
    "test <- filtered_test[ , c(5,6,11,13,16,35,36,37,38,39)]\n",
    "colnames(test)[1] <- \"#CHROM\"\n",
    "colnames(test)[2] <- \"POS\"\n",
    "colnames(test)[3] <- \"REF\"\n",
    "colnames(test)[4] <- \"ALT\"\n",
    "colnames(test)[5] <- \"Patient_ID\"\n",
    "write.table(test, file = \"~/R/LTRC_data/patient_reference_file\", sep = \"\\t\", quote=FALSE, row.names=FALSE )\n",
    "\n",
    "\n",
    "#merge variants and patient reference files\n",
    "variants <- read.delim(\"~/R/LTRC_data/variants\", sep = \"\\t\", header = T)\n",
    "patient_reference_file <- read.delim(\"~/R/LTRC_data/patient_reference_file\", sep = \"\\t\", header = T)\n",
    "#library(dplyr)\n",
    "#count <- patient_reference_file%>%count(\"Patient_ID\")\n",
    "#count1 <- test%>%count(\"Patient_ID\")\n",
    "\n",
    "filtered_variants_final <- patient_reference_file %>% left_join(variants)\n",
    "write.table(filtered_variants_final, file = \"~/R/LTRC_data/filtered_variants_final\", sep = \"\\t\", quote=FALSE, row.names=FALSE )\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## filter out variants occurring in more than 10 individuals\n",
    "#import data\n",
    "filtered_variants_final <- read.delim(\"~/R/LTRC_data/filtered_variants_final\", sep = \"\\t\", header = T)\n",
    "library(dplyr)\n",
    "variants_filtered_10 <- filtered_variants_final %>%\n",
    "  group_by(X.CHROM,POS,REF,ALT) %>%\n",
    "  distinct(.keep_all = TRUE) %>%\n",
    "  filter(n() <= 10)\n",
    "write.table(variants_filtered_10,\"~/R/LTRC_data/variants_filtered_10\", sep = \"\\t\", quote=FALSE, row.names=FALSE ) # 284,997 variants; 1346 patients\n",
    "#file\n",
    "variants_filtered_10 <- read.delim(\"~/R/LTRC_data/variants_filtered_10\", sep = \"\\t\", header = T)\n",
    "\n",
    "\n",
    "\n",
    "#double check count\n",
    "#library(plyr)\n",
    "#count <- ddply(variants_filtered_10,.(X.CHROM,POS,REF,ALT),nrow)\n",
    "#count <- variants_filtered_10%>%count(\"Patient_ID\")\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.0"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
